File: concepts/03_errors_and_resources/02_exception_propagation_and_matching/r/test_02_rethrow_and_completion_precedence.R
# polyglot-family: errors_and_resources
# polyglot-concept: exception_propagation_and_matching
# polyglot-related: languages/r/language/11_conditions_control_flow_and_cleanup/
# polyglot-related+: test_087_nested_failures_and_condition_context.R
#
# 共同问题：捕获后怎样保留原错误重新传播；清理失败与原失败谁胜出。
# 对照观察：handler 中再次 `stop(condition)` 传播同一对象；`on.exit` 的错误可覆盖正在传播的错误。

original <- simpleError("original")
rethrown <- tryCatch(
    tryCatch(stop(original), error = function(condition) stop(condition)),
    error = identity
)
fail_during_cleanup <- function() {
    on.exit(stop("cleanup"), add = TRUE)
    stop("body")
}
cleanup_wins <- tryCatch(fail_during_cleanup(), error = identity)

stopifnot(
    identical(rethrown, original),
    identical(conditionMessage(cleanup_wins), "cleanup")
)

File: concepts/03_errors_and_resources/02_exception_propagation_and_matching/rust/test_02_rethrow_and_completion_precedence.rs
// polyglot-family: errors_and_resources
// polyglot-concept: exception_propagation_and_matching
// polyglot-related: languages/rust/tests/language/08_errors_panics_and_resource_management/
// polyglot-related+: test_060_panic_catch_unwind_and_unwind_safety.rs
//
// 共同问题：重新传播是否保留原失败；清理期间再次失败时哪个 completion 胜出。
// 对照观察：`resume_unwind` 保留原 panic payload；Drop 再 panic 会导致 abort，因此不得用它替代错误聚合。

#[test]
fn comparison() {
    let outer = std::panic::catch_unwind(|| {
        let payload = std::panic::catch_unwind(|| std::panic::panic_any(String::from("original")))
            .unwrap_err();
        std::panic::resume_unwind(payload);
    });
    let payload = outer.unwrap_err();
    assert_eq!(
        payload.downcast_ref::<String>().map(String::as_str),
        Some("original")
    );
}
